### PR TITLE
Feat/script

### PR DESCRIPTION
**Pull request by Najo-z**

### CODE REVIEW

File: libscript/include/pivot/script/Exceptions.hxx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+#include <string>
+
+namespace pivot::ecs::script
+{
+
+class MixedIndentException : public std::runtime_error
+{
+public:
+    MixedIndentException(const char *msg, const char *line_, size_t linenb_, const char *info_)
+        : std::runtime_error(msg), line(line_), linenb(linenb_), info(info_)
```
Comment (zcorniere) [inline libscript/include/pivot/script/Exceptions.hxx:13]:
```suggestion
        : std::runtime_error(msg), linenb(linenb_),  line(line_), info(info_)
```
Initialization must occur in the same order as the member variable declaration.

File: libscript/include/pivot/script/Exceptions.hxx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    size_t get_line_nb() const { return linenb; }
+    /// return the error raw line
+    const char *get_line() const { return line; }
+    /// return the error specification
+    const char *get_info() const { return info; }
+
+private:
+    size_t linenb;
+    const char *line;
+    const char *info;
```
Comment (zcorniere) [inline libscript/include/pivot/script/Exceptions.hxx:26]: UNSAFE 
if you can't be sure that the input char * would be static, you must not store raw, unmanaged pointer.

Line can be a int
and info should be in the message

You can also store your char * as a std::runtime_error.
```suggestion
    size_t get_line_nb() const { return linenb; }
    /// return the error raw line
    const char *get_line() const { return line.what(); }
    /// return the error specification
    const char *get_info() const { return info.what(); }

private:
    size_t linenb;
    std::runtime_error line;
    std::runtime_error info;
```
It is a weird but valid use case. Still, it would be best if you could limit its usage.

File: libscript/include/pivot/script/Exceptions.hxx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    const char *getExceptionType() const { return exceptionType; }
+    /// return the representation of what caused the error
+    const char *getFaulter() const { return faulter; }
+    /// return the error specification
+    const char *getError() const { return error; }
+
+private:
+    const char *exceptionType;
+    const char *faulter;
+    const char *error;
```
Comment (zcorniere) [inline libscript/include/pivot/script/Exceptions.hxx:78]: UNSAFE
same as above

File: libscript/include/pivot/script/Parser.hxx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+void consumeSystemFuncParams(std::vector<Token> &tokens, Node &result,
+                             Token &lastToken);    // consume function parameters and append them to children node
+
+void expectSystemToken(
+    std::vector<Token> &tokens, TokenType expectedType, Token &lastToken,
+    bool consume);    // check that token exists, and is of correct type (and potentially consume it from tokens)
+void expectSystemTokenValue(
+    std::vector<Token> &tokens, const std::string &expectedValue, Token &lastToken,
+    bool consume);    // check that token exists, and is of correct value (and potentially consume it from tokens)
+bool isDeclarationOver(const std::vector<Token> &tokens);    // Check for end of system/component declaration
```
Comment (zcorniere) [inline libscript/include/pivot/script/Parser.hxx:42]: Could you wrap all of that into a function with method ?

You almost always carry around a `std::vector<Token>`, that could be a member variable, and everything else could be method.

File: libscript/include/pivot/script/Stack.hxx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace pivot::ecs::script::interpreter
+{
+
+struct Variable {
+    std::string name;
+    bool hasValue = false;
+    data::Value value;
+    std::unordered_map<std::string, Variable> members;
+};
```
Comment (zcorniere) [inline libscript/include/pivot/script/Stack.hxx:16]: Is considered by gcc as a forward declaration, no idea why
@Ersikan

File: libscript/include/pivot/script/Stack.hxx [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * \brief	Container for all the variables a system
+ * 			holds during the execution of a tick
+ *
+ * 	This container allows access, use and modification of the
+ * 	variables it stores. This is use by the interpreter to handle
+ * 	variables during the execution of a tick.
+ */
+class Stack
```
Comment (zcorniere) [inline libscript/include/pivot/script/Stack.hxx:26]: Could be a template class ?
```cpp
template<typename Key, typename Value>
```

File: libscript/sources/Builtins.cxx [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+#include "pivot/script/Exceptions.hxx"
+
+#include <iostream>
+
+namespace pivot::ecs::script::interpreter::builtins
+{
+
+// Builtins assume that the interpreter has called interpreter::validateParams()
+// therefore that the parameters are of correct size and types
+
```
Comment (zcorniere) [inline libscript/sources/Builtins.cxx:12]: The doc must be in the header :)

File: libscript/sources/Builtins.cxx [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+    return data::Value(true);
+}
+
+// void	print(String/Number/Boolean param1, ...)
+//		print the parameters on a single line, ending with a newline
+data::Value builtin_print(const std::vector<data::Value> &params)
+{
+    for (const data::Value &param: params) {    // print has unlimited number of parameters
+        data::BasicType varType = std::get<data::BasicType>(param.type());
+        if (varType == data::BasicType::Number)
```
Comment (zcorniere) [inline libscript/sources/Builtins.cxx:27]: switch case ?

File: libscript/sources/Builtins.cxx [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+            return data::Value(std::get<bool>(left) == std::get<bool>(right));    // perform arithmetic comparison
+        else if (std::get<data::BasicType>(left.type()) == data::BasicType::Vec3 &&
+                 std::get<data::BasicType>(right.type()) == data::BasicType::Vec3)
+            return data::Value(std::get<glm::vec3>(left) == std::get<glm::vec3>(right));    // let glm do the comparison
+        else                                                                                // unsupported
+            throw InvalidException("InvalidOperation", (left.type().toString() + "," + right.type().toString()).c_str(),
+                                   "Invalid equal to '==' operator between these types.");
+    } catch (InvalidException e) {
+        throw e;
+    } catch (std::bad_variant_access e) {
```
Comment (zcorniere) [inline libscript/sources/Builtins.cxx:64]: You MUST catch exception as a const ref

File: libscript/sources/Builtins.cxx [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+                 std::get<data::BasicType>(right.type()) == data::BasicType::Boolean)
+            return data::Value(std::get<bool>(left) == std::get<bool>(right));    // perform arithmetic comparison
+        else if (std::get<data::BasicType>(left.type()) == data::BasicType::Vec3 &&
+                 std::get<data::BasicType>(right.type()) == data::BasicType::Vec3)
+            return data::Value(std::get<glm::vec3>(left) == std::get<glm::vec3>(right));    // let glm do the comparison
+        else                                                                                // unsupported
+            throw InvalidException("InvalidOperation", (left.type().toString() + "," + right.type().toString()).c_str(),
+                                   "Invalid equal to '==' operator between these types.");
+    } catch (InvalidException e) {
+        throw e;
```
Comment (zcorniere) [inline libscript/sources/Builtins.cxx:63]: Why catching an exception just to rethrow it. Remove

File: libscript/sources/Builtins.cxx [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+    } catch (std::bad_variant_access e) {
+        throw InvalidException("InvalidOperation", (left.type().toString() + "," + right.type().toString()).c_str(),
+                               "Invalid not equal to '!=' operator between these types.");
+    }
+}
+data::Value builtin_operatorGt(const data::Value &left, const data::Value &right)
+{    // greater than operator '>'
+    try {
+        if (std::get<data::BasicType>(left.type()) == data::BasicType::Number &&
+            std::get<data::BasicType>(right.type()) == data::BasicType::Number)
```
Comment (zcorniere) [inline libscript/sources/Builtins.cxx:99]: You could use [std::variant::holds_alternative](https://en.cppreference.com/w/cpp/utility/variant/holds_alternative)

File: libscript/sources/Builtins.cxx [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+            return data::Value(std::get<double>(left) > std::get<double>(right));    // perform arithmetic comparison
+        else if (std::get<data::BasicType>(left.type()) == data::BasicType::String &&
+                 std::get<data::BasicType>(right.type()) == data::BasicType::String)
+            return data::Value(std::get<std::string>(left) >
+                               std::get<std::string>(right));    // perform alphabetical comparison
```
Comment (zcorniere) [inline libscript/sources/Builtins.cxx:104]: If you use `std::get_if<std::string>` it would return nullptr if the conversion fail, avoiding the try / catch block just to retrow an exception.

File: libscript/sources/Engine.cxx [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+{
+
+Engine::Engine(systems::Index &systemIndex, component::Index &componentIndex)
+    : _systemIndex(systemIndex), _componentIndex(componentIndex)
+{
+}
+
+std::string Engine::loadFile(const std::string &file, bool isContent, bool verbose)
+{    // Load a file to register all descriptions in the file, as well as store the trees for the system entry points
+    if (verbose) std::cout << "Loading file " << file << std::endl;
```
Comment (zcorniere) [inline libscript/sources/Engine.cxx:15]:
```suggestion
    if (verbose) logger.info(Script Engine) << "Loading file " << file << std::endl;
```

File: libscript/sources/Engine.cxx [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        desc.system = std::bind(&Engine::systemCallback, this, std::placeholders::_1, std::placeholders::_2,
+                                std::placeholders::_3);    // add the callback to the descriptions
+        _systemIndex.registerSystem(desc);                 // register system in the not-so-global index
+        try {
+            Node systemEntry = getEntryPointFor(
+                desc.name, fileNode);    // only store the entry point for the system, the rest is in the description
+            _systems.insert({desc.name, systemEntry});
+        } catch (std::exception e) {
+            return e.what();    // shouldn't happen
+        }
```
Comment (zcorniere) [inline libscript/sources/Engine.cxx:33]: Ha ha, that funny.

Prefer returning a bool

```suggestion
        } catch (std::exception e) {
            logger.err("Engine::loadFile") << e.what();    // will happen
            return false;
        }
```

File: libscript/sources/Engine.cxx [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    }
+    return (file + " parsed succesfully.");    // no errors
```
Comment (zcorniere) [inline libscript/sources/Engine.cxx:35]:
```suggestion
    return true;
```

File: libscript/sources/Engine.cxx [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        // c++20 gcc yet
+        logger.err("Unregistered system '") << system.name << "'";
+        return;
+    }
+    const Node &systemEntry = _systems.at(system.name);    // Avoid looking up for every entity
+    for (auto entity: entities) {                          // For every entity, execute the system with it as parameter
+        try {
+            _stack.clear();
+            interpreter::executeSystem(systemEntry, system, entity, trigger, _stack);
+        } catch (InvalidException e) {
```
Comment (zcorniere) [inline libscript/sources/Engine.cxx:52]: Exception by const reference pls

File: libscript/sources/Engine.cxx [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+            logger.warn(e.getExceptionType()) << " '" << e.getFaulter() << "' => " << e.getError();
```
Comment (zcorniere) [inline libscript/sources/Engine.cxx:53]: I see why you want to have multiple char * value.
For now, you can use the `std::runtime_error` technique I mentioned earlier, but take a lot at the file "PivotException.hxx" in the lib graphics. It is more or less the same needs

File: libscript/sources/Engine.cxx [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+
+Node Engine::getEntryPointFor(const std::string &systemName, const Node &file)
+{    // Get the entry point for a system name in a file root tree node
+    for (const Node &declaration: file.children)
+        if (declaration.type == NodeType::SystemDeclaration && declaration.value == systemName)
+            for (const Node &child: declaration.children)
+                if (child.type == NodeType::SystemEntryPoint) return child;
+    // throw(std::exception(std::format("No entry point found for {}", systemName).c_str())); // format not available in
+    // c++20 gcc yet
+    throw(std::runtime_error(("No entry point found for " + systemName).c_str()));
```
Comment (zcorniere) [inline libscript/sources/Engine.cxx:70]: throw does not need parenthesis

File: libscript/sources/Interpreter.cxx [hunk 19]
```diff
@@ -0,0 +1,531 @@
...
+#include "Logger.hpp"
+#include "pivot/script/Builtins.hxx"
+#include "pivot/script/Exceptions.hxx"
+#include <limits>
+#include <unordered_map>
+
+namespace pivot::ecs::script::interpreter
+{
+
+// TODO: consider using magicenum
```
Comment (zcorniere) [inline libscript/sources/Interpreter.cxx:11]: yep

File: libscript/sources/Interpreter.cxx [hunk 20]
```diff
@@ -0,0 +1,531 @@
...
+    {"!=", interpreter::builtins::builtin_operatorNEq},  {"&&", interpreter::builtins::builtin_operatorAnd},
+    {"||", interpreter::builtins::builtin_operatorOr}};
+// Map built-in function strings, to their callback and parameters (number and types)
+const std::unordered_map<              // map
+    std::string, std::pair<            // function string name, to a pair
+                     std::function<    // first : callback
+                         data::Value(const std::vector<data::Value> &)>,
+                     std::pair<size_t, std::vector<std::vector<data::Type>>>    // second: pair; first: number of
+                                                                                // params, second: types of params
+                     >>
```
Comment (zcorniere) [inline libscript/sources/Interpreter.cxx:48]: shesh, that is a complicated type

consider using something like this for clarity
```suggestion
/// doc
using FunctionCallback = std::function<data::Value(const std::vector<data::Value> &)>
/// Doc
using ParameterPair = std::pair<size_t, std::vector<std::vector<data::Type>>>

const std::unordered_map<              // map
    std::string, std::pair<FunctionCallback, ParameterPair>>
```

File: libscript/sources/Interpreter.cxx [hunk 21]
```diff
@@ -0,0 +1,531 @@
...
+        for (const Node &node: file.children) {
+            if (node.type == NodeType::ComponentDeclaration) {    // register component
+                registerComponentDeclaration(node, componentIndex, file.value);
+            } else if (node.type == NodeType::SystemDeclaration) {    // store system declaration for return
+                systems::Description r = registerSystemDeclaration(node, file.value);
+                if (r.name == "Error 1") {
+                    // std::cerr << std::format("registerDeclarations(const Node &file): failed to interpret sub node
+                    // {}", gNodeTypeStrings.at(node.type)) << std::endl; // format not available in c++20 gcc yet
+                    logger.err("registerDeclarations(const Node &file): failed to register declarations for sub node ")
+                        << gNodeTypeStrings.at(node.type);
```
Comment (zcorniere) [inline libscript/sources/Interpreter.cxx:79]: The argument of the logger call is supposed to be a short word :) 
```suggestion
                    logger.err("Register Declarations") << "const Node &file: failed to register declarations for sub node :"
                        << gNodeTypeStrings.at(node.type);
```

File: libscript/sources/Parser.cxx [hunk 22]
```diff
@@ -0,0 +1,880 @@
...
+            else    // neither an identifier, a component nor a system
+                throw TokenException("ERROR", t.value.c_str(), t.line_nb, t.char_nb, "UnexpectedToken",
+                                     "Expected 'component' or 'system' token");
+        }
+        if (verbose) printFileNode(result);
+    } catch (MixedIndentException e) {
+        // std::cerr << std::format("\nParser {} {}:\t{}\n\tline {}:\t{}\n{}",
+        // 	e.what(), "IndentException" , file, e.get_line_nb(), e.get_line(), e.get_info()) << std::endl; //
+        // format not available in c++20 gcc yet
+        logger.err("\nParser ") << e.what() << " IndentException "
```
Comment (zcorniere) [inline libscript/sources/Parser.cxx:90]: why `\n` at the begining of the logger message ? remove it

File: libscript/sources/Parser.cxx [hunk 23]
```diff
@@ -0,0 +1,880 @@
...
+        lineEnd = text.find('\n', lineStart);
+        lineNb++;
+    }
+
+    return result;
+}
+
+bool isDeclarationOver(const std::vector<Token> &tokens)
+{                              // Is there no more tokens, or is the first the end of the declaration
+    if (tokens.size() == 0)    // No more tokens
```
Comment (zcorniere) [inline libscript/sources/Parser.cxx:724]:
```suggestion
    if (tokens.empty())    // No more tokens
```

File: libscript/sources/Parser.cxx [hunk 24]
```diff
@@ -0,0 +1,880 @@
...
+
+#include "Logger.hpp"
+#include "pivot/script/Exceptions.hxx"
+#include "pivot/script/Parser.hxx"
+
+namespace pivot::ecs::script::parser
+{
+
+const std::string gKnownSymbols = "+-/*%=!()<>,.#&|\" \t\r\n";
+const std::string gWhitespace = " \t\r\n";
```
Comment (zcorniere) [inline libscript/sources/Parser.cxx:16]: Please use [std::isspace](https://en.cppreference.com/w/cpp/string/byte/isspace)

File: libscript/tests/tests/lexer/OneLineComponent.pvt [hunk 25]
```diff
@@ -0,0 +1 @@
+component Test = Number
```
Comment (zcorniere) [inline libscript/tests/tests/lexer/OneLineComponent.pvt:1]: didn't you added this script to the test ?

remove it if possible

File: libscript/tests/tests/physics.pvt [hunk 26]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (zcorniere) [inline libscript/tests/tests/physics.pvt:1]: didn't you added this script to the test ?

remove it if possible

File: libscript/include/pivot/script/DataStructures.hxx [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+    size_t line_nb;
+    size_t char_nb;
+
+    // inline std::string toString() const { return std::format("value='{}' line {} char {}", value, line_nb, char_nb);
+    // } // format not available in c++20 gcc yet
+    inline std::string toString() const
+    {
+        return "value='" + value + "' line " + std::to_string(line_nb) + " char " + std::to_string(line_nb);
+    }
+    inline bool operator==(Token rhs) const { return tied() == rhs.tied(); }
```
[the remark this answers is not in the document]
Comment (Najo-z) [reply libscript/include/pivot/script/DataStructures.hxx:101]: That operator cannot be defaulted as suggested in the suggested change.
However, applying your change as well as making the Token parameter a const reference works.

File: libscript/include/pivot/script/Stack.hxx (hunk 6 above)
Comment (Najo-z) [reply libscript/include/pivot/script/Stack.hxx:26]: Usage is pretty specific, and requires recursion in the Value field (since it searches that way inside the container). Therefore, I don't think there's enough other use cases for it to be templated.

File: libscript/sources/Builtins.cxx (hunk 11 above)
Comment (Najo-z) [reply libscript/sources/Builtins.cxx]: I'm using the try/catch because any single fail in std::get (caused by unexpected type) should result in stopping execution and unwinding to the top (by throwing a custom exception describing the error).
I could indeed use ifs to check if std::get would fail, and then throw, but I feel like it bloats the code.
What do you think ?

File: libscript/sources/Builtins.cxx (hunk 12 above)
Comment (Najo-z) [reply libscript/sources/Builtins.cxx]: I'm using the try/catch because any single fail in std::get (caused by unexpected type) should result in stopping execution and unwinding to the top (by throwing a custom exception describing the error).
I could indeed use ifs to check if std::get would fail, and then throw, but I feel like it bloats the code.
What do you think ?

File: libscript/include/pivot/script/Stack.hxx (hunk 5 above)
Comment (Najo-z) [reply libscript/include/pivot/script/Stack.hxx]: So we understood that it's because of `std::unordered_map<std::string, Variable>` which is a pretty nooby mistake by me.
Since you may only store pointers or references, I'm guessing I'll be using `std::unique_ptr<Variable>`.
This should be fine, right ?

File: libecs/include/pivot/ecs/Core/Systems/description.hxx [hunk 28]
```diff
@@ -21,6 +21,8 @@ struct Description {
 
     /// system name
     std::string name;
+    /// entity name
+    std::string entityName;
```
Comment (Ersikan) [inline libecs/include/pivot/ecs/Core/Systems/description.hxx:25]: Which entities name is it ? Why did you add only one name, even though the system can work on multiple entities ?

File: libscript/sources/Builtins.cxx [hunk 29]
```diff
@@ -0,0 +1,325 @@
...
+                 std::get<data::BasicType>(right.type()) == data::BasicType::Vec3)
+            return data::Value(std::get<glm::vec3>(left) == std::get<glm::vec3>(right));    // let glm do the comparison
+        else {                                                                              // unsupported
+            logger.err("ERROR") << " by '" << left.type().toString() << "' and '" << right.type().toString() << "'";
+            throw InvalidOperation("Invalid equal to '==' operator between these types.");
+        }
+    } catch (const std::bad_variant_access &) {
+        logger.err("ERROR") << " by '" << left.type().toString() << "' and '" << right.type().toString() << "'";
+        throw InvalidOperation("Invalid equal to '==' operator between these types.");
+    }
```
Comment (Ersikan) [inline libscript/sources/Builtins.cxx:60]: Please use `std::visit` to write generically all this code.

File: libscript/include/pivot/script/DataStructures.hxx [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+/// An expression op is either an operator or an operand
+/// A chain of these represents an entire expression (stored in postfix)
+struct ExpressionOp {
+    /// Whether this is an operator or an operand
+    bool isOperator;
+    /// Value of the operand
+    data::Value operand;
+    /// String representing the operator
+    std::string operatorStr;
+};
```
Comment (Ersikan) [inline libscript/include/pivot/script/DataStructures.hxx:79]: You could use an `std::variant`.
```suggestion
using Operand = data::Value;
using Operator = std::string;
using ExpressionOp = std::variant<Operand, Operator>;
```

File: libscript/sources/Parser.cxx [hunk 31]
```diff
@@ -0,0 +1,930 @@
...
+    } catch (const UnexpectedTokenValueException &e) {
+        logger.err("Unexpected Token Value: ") << e.what();
+    } catch (const std::invalid_argument &e) {    // logic error
+        logger.err("LogicError: ") << e.what();
+    } catch (const std::exception &e) {
+        // std::cerr << std::format("\nParser !Unhandled Exception!: {}", e.what()) << std::flush; // format not
+        // available in c++20 gcc yet
+        logger.err("\nscript::parser !Unhandled Exception!: ") << e.what();
+    }
+    return result;
```
Comment (Ersikan) [inline libscript/sources/Parser.cxx:80]: Do not return an incomplete node when an exception occured.

File: libscript/include/pivot/script/Parser.hxx [hunk 32]
```diff
@@ -0,0 +1,89 @@
...
+ * 	and generate an AST based on syntax.
+ *
+ */
+class Parser
+{
+public:
+    ///	Main function, will return the abstract syntax tree of the file (root Node)
+    Node ast_from_file(const std::string &file, bool isContent = false, bool verbose = true);
+    /// Lexer function, will return a list of Tokens contained in the file
+    void tokens_from_file(const std::string &file, bool isContent = false, bool verbose = false);
```
Comment (Ersikan) [inline libscript/include/pivot/script/Parser.hxx:23]: Take an istream and remove `isContent`
```suggestion
    ///	Main function, will return the abstract syntax tree of the file (root Node)
    Node ast_from_file(std::istream input, bool verbose = true);
    /// Lexer function, will return a list of Tokens contained in the file
    void tokens_from_file(std::istream input, bool verbose = false);
```

File: libscript/include/pivot/script/Stack.hxx [hunk 33]
```diff
@@ -0,0 +1,55 @@
...
+struct Variable {
+    /// String representing the name of the variable
+    std::string name;
+    /// Whether the variable has a value or not
+    bool hasValue = false;
+    /// The value of the variable
+    data::Value value;
+    /// A map representing its members/fields (also variables)
+    std::unordered_map<std::string, Variable> members;
+};
```
Comment (Ersikan) [inline libscript/include/pivot/script/Stack.hxx:24]: Replace with `data::Value`.

File: libscript/sources/Engine.cxx [hunk 34]
```diff
@@ -0,0 +1,77 @@
...
+        desc.system = std::bind(&Engine::systemCallback, this, std::placeholders::_1, std::placeholders::_2,
+                                std::placeholders::_3);    // add the callback to the descriptions
+        _systemIndex.registerSystem(desc);                 // register system in the not-so-global index
+        try {
+            Node systemEntry = getEntryPointFor(
+                desc.name, fileNode);    // only store the entry point for the system, the rest is in the description
+            _systems.insert({desc.name, systemEntry});
+        } catch (const std::exception &e) {
+            return e.what();    // unlikely branch, return error to dearImGui
+        }
```
Comment (Ersikan) [inline libscript/sources/Engine.cxx:33]: Do not catch the error here. If the error must be caught to display in DearImGui, do it in the editor.
```suggestion
        Node systemEntry = getEntryPointFor(
            desc.name, fileNode);    // only store the entry point for the system, the rest is in the description
        _systems.insert({desc.name, systemEntry});
```

File: libscript/tests/test_interpreter.cxx [hunk 35]
```diff
@@ -0,0 +1,95 @@
...
+    component::ArrayCombination combinations{{std::ref(*array1), std::ref(*array2)}};
+    event::EventWithComponent evt = {
+        .event = event::Event{.description = sysdescription.eventListener, .entities = {1, 2}, .payload = 0.12},
+        .components = {{// mdr<Position, Velocity> event parameter
+                        component::ComponentRef(*array1, 1), component::ComponentRef(*array2, 1)},
+                       {// lol<Velocity> event parameter
+                        component::ComponentRef(*array2, 2)}}};
+
+    sysdescription.system(sysdescription, combinations, evt);
+
```
Comment (Ersikan) [inline libscript/tests/test_interpreter.cxx:93]: You don't interpret or check anything in this test ?

**Review by Ersikan — CHANGES_REQUESTED** [review]

File: libecs/include/pivot/ecs/Core/Event/description.hxx [hunk 36]
```diff
@@ -19,6 +19,8 @@ struct Description {
     std::vector<std::string> entities;
     /// Value of the payload
     data::Type payload;
+    /// Name of the payload
+    std::string payloadName;
```
[the remark this answers is not in the document]
Comment (Najo-z) [reply libecs/include/pivot/ecs/Core/Event/description.hxx:23]: A payload has a type and a name. The name was not stored in the description, but I see no reason not to. All the information about a system/component declaration are in the description, this should be as well.
Also this means the engine does not need to store tiny bits of information (such as this) itself, rather put it in the description.

File: libecs/include/pivot/ecs/Core/Systems/description.hxx (hunk 28 above)
Comment (Najo-z) [reply libecs/include/pivot/ecs/Core/Systems/description.hxx:25]: Actually, it doesn't.
IIRC, a system may only take a single entity as parameter for now, as suggests this bit of code, in ecs::systems::Description.
`    std::vector<std::string> systemComponents;
`
Which is a vector representing the components THE (only) entity parameter has.
@Maxteke

File: libscript/include/pivot/script/Parser.hxx (hunk 32 above)
Comment (Najo-z) [reply libscript/include/pivot/script/Parser.hxx:23]: This would allow using a single parameter which could act both as a file, or a harcoded string, which is exactly what I was looking for.
However, I don't understand how to differentiate that once my function has been called.
How can i know the file name (to display to the user in case of error for example) or distinguish between file/string if the stream has been constructed before the call to the function ?

File: libecs/include/pivot/ecs/Core/Systems/description.hxx (hunk 28 above)
Comment (Ersikan) [reply libecs/include/pivot/ecs/Core/Systems/description.hxx:25]: Yes, the system runs on only one entity, but the event itself can provide some entities
Anyway, why include the entity name, which is just the name used to interact with it *inside* the system ? When *using* the system, the name of the entity is not useful.

File: libscript/include/pivot/script/Parser.hxx (hunk 32 above)
Comment (Ersikan) [reply libscript/include/pivot/script/Parser.hxx:23]: For giving context to the user, you could add the file name as a parameter, with some default value.
To read from the stream, you don't need to know whether the underlying source is a string, a file, or anything else. You never need to distinguish between the two.

File: libscript/include/pivot/script/Stack.hxx (hunk 5 above)
Comment (Najo-z) [reply libscript/include/pivot/script/Stack.hxx]: After the review from @Ersikan , I just removed it since it was a useless re-implementation of pivot::ecs::data::Value.

**Review by Ersikan — APPROVED** [review]

**Review by zcorniere — APPROVED** [review]
